NuRadioReco/modules/io/coreas/readCoREASDetector.py

```diff
@@ -0,0 +1,330 @@
+    """
+    observer = np.zeros((efield_times.shape[0], 4))
+    observer[:, 0] = efield_times
+    observer[:, 1:4] = efields
+    observer = observer.T
+    return observer
+
+def select_channels_per_station(det, station_id, requested_channel_ids):
+    """
+    Returns a defaultdict object containing the requested channel ids that are in the given station.
+    This dict contains the channel group ids as keys with lists of channel ids as values.
+
+    Parameters
+    ----------
+    det : DetectorBase
+        The detector object that contains the station
+    station_id : int
+        The station id to select channels from
+    requested_channel_ids : list
+        List of requested channel ids
+    """
+    channel_ids = defaultdict(list)
+    for channel_id in requested_channel_ids:
+        if channel_id in det.get_channel_ids(station_id):
+            channel_group_id = det.get_channel_group_id(station_id, channel_id)
+            channel_ids[channel_group_id].append(channel_id)
+    return channel_ids
+
+class readCoREASDetector:
+    """
+    Use this as default when reading CoREAS files and combining them with a detector.
+
+    This model reads the electric fields of a CoREAS file with a star shaped pattern and foldes them with it given detector. 
+    The core position of the star shape pattern is randomly distributed within a user defined area. 
+    If interpolated=True the electric field of the star shaped pattern is interpolated at the detector positions. 
+    If interpolated=False the closest observer position of the star shape pattern to the detector is chosen.
+    """
+
+    def __init__(self):
+        self.__t = 0
+        self.__t_event_structure = 0
+        self.__t_per_event = 0
+        self.__input_files = None
+        self.__n_cores = None
+        self.__current_input_file = None
+        self.__random_generator = None
+        self.__interp_lowfreq = None
+        self.__interp_highfreq = None
+        self.logger = logging.getLogger('NuRadioReco.readCoREASDetector')
+
+    def begin(self, input_files, xmin=0, xmax=0, ymin=0, ymax=0, n_cores=10, seed=None, core_position_list=[], selected_station_ids=[], selected_channel_ids=[], interp_lowfreq=30*units.MHz, interp_highfreq=1000*units.MHz, log_level=logging.INFO, debug=False):
+            
+        """
+        begin method
+        initialize readCoREAS module
+        Parameters
+        ----------
+        input_files: input files
+            list of coreas hdf5 files
+        xmin: float
+            minimum x coordinate of the area in which core positions are distributed
+        xmax: float
+            maximum x coordinate of the area in which core positions are distributed
+        ymin: float
+            minimum y coordinate of the area in which core positions are distributed
+        ynax: float
+            maximum y coordinate of the area in which core positions are distributed
+        n_cores: number of cores (integer)
```

Comment:
The readCoreasDetector should also replace the readCoreasStationGrid. Since the name is misleading I started a new module instead of applying the changes there. How it is currently implemented should allow backwards compatibility. But maybe that's not necessary because of the new name? @cg-laser , do you have any opinions about this?

Reply:
I am not sure that I agree here. The benefit of having "randomness" in the module is, that (when done correctly) you allow to reproduce random results (by defining the random generator and fixing the seed) which is important for reproducibility, validation and so on.

Reply:
@fschlueter I am not sure I understand your comment. My suggestion would be to always pass a list of shower cores to the run() function. This list can easily be generated with a random generator, to keep the random behavior in cases where it is needed. But I would like this to be handled outside of the module, because it obscures the functionality of the module itself. What is benefit of handling the randomness inside the module?

Reply:
It is easier to control the randomness when it is within the module. And I do not see why it would obscure the functionality, can you elaborate. However, I can foresee a use case where you want to pass a non-random list of cores to the module. Hence, I do not have strong feelings about either solution if it is ensured that the randomness is reproducible easily.